keyboards/annepro2/protocol.h

```diff
@@ -0,0 +1,113 @@
+/*
+ * (c) 2021 by Tomasz bla Fortuna
+ * License: GPLv2
+ *
+ * This file is shared with Shine firmware.
+ */
+
+#ifndef PROTOCOL_INCLUDED
+#define PROTOCOL_INCLUDED
+#include <inttypes.h>
+
+#define PROTOCOL_SD SD0
+
+enum {
+  /*
+   * Main -> LED
+   */
+  /* Basic config */
+  CMD_LED_ON = 0x01,
+  CMD_LED_OFF = 0x02,
+
+  CMD_LED_SET_PROFILE = 0x03,
+  CMD_LED_NEXT_PROFILE = 0x04,
+  CMD_LED_PREV_PROFILE = 0x05,
+
+  CMD_LED_NEXT_INTENSITY = 0x06,
+  CMD_LED_NEXT_ANIMATION_SPEED = 0x07,
+
+  /* Masks */
+  /* Override a key color, eg. capslock */
+  CMD_LED_MASK_SET_KEY = 0x10,
+  /* Override all keys in a row with configurable colors */
+  CMD_LED_MASK_SET_ROW = 0x11,
+
+  /* Override all keys with single color (eg. foreground color) */
+  CMD_LED_MASK_SET_MONO = 0x12,
+
+  /* Reactive / status */
+  CMD_LED_GET_STATUS = 0x20,
+  CMD_LED_KEY_BLINK = 0x21, /* TODO: Blink */
+  CMD_LED_KEY_DOWN = 0x22,
+  CMD_LED_KEY_UP = 0x23, /* TODO */
+  CMD_LED_IAP = 0x24,
+
+  /* LED -> Main */
+  /* TODO: Payload with data to send over HID */
+  CMD_LED_DEBUG = 0x40,
+
+  /* Number of profiles, current profile, on/off state,
+     reactive flag, brightness, errors */
+  CMD_LED_STATUS = 0x41,
+};
+
+/* 4 ROWS * 14 COLS * 4B (RGBX) = 224 */
+#define MAX_PAYLOAD_SIZE 230
+
+enum protoState {
+  /* 2-byte initial start-of-message sync */
+  FSA_SYNC_1,
+  FSA_SYNC_2,
+  /* Waiting for command byte */
+  FSA_CMD,
+  /* Waiting for ID byte */
+  FSA_ID,
+  /* Waiting for payload size */
+  FSA_PAYLOAD_SIZE,
+  /* Reading payload until payloadPosition == payloadSize */
+  FSA_PAYLOAD,
+};
+
+/* Buffer holding a single message */
+typedef struct {
```

Comment:
I am not sure if this is the best idea. I understand that stuff like this can be allocated on the stack, but even then we are still quite limited in RAM. Consider on C15? with the HT32F52342 we have only 8K of ram. (We can take a look at the static ram usage from objdump, but just something to keep in mind).

One potential optimization would be if we swap the payloadSize and payload fields, then we can _potentially_ allocate a smaller buffer than this struct. It has problems of its own, ie. in theory it should work, but it is really prone to error. 

I would say this is not a problem so per say, but just something we should consider / discuss before committing to something.

Reply:
It's possible to do header sent in one request and then payload in another one - without copying the payload. I haven't choose to do so because of a single reason:  considering a message buffering and ACKing of a message + resend if ack doesn't come. To resend I need a complete message.

With a single call it's possible to do a blind-retry without copying too.

And... maybe just retrying 3 times blindly by default is not that stupid. It seems to work very nicely on my board. Unsure about more broken ones. If there's no need for resend+ack then the payload can be dropped from the message.

Reply:
I confused receiving and transmitting in previous message. Once again:
1) Receiving has to store parts of the message somewhere, but the buffer can be much smaller. I calculated it for something much bigger than is transmitted.
2) Transmitting can use separate calls to transmit payload, but it needs more than one call as data is usually padded, ie. row number and then row data. Ie. if splitting transmission into multiple calls, then automatic retries won't be easy. But, those make sense only for short messages - so I can do probably duplicated API. One for short messages with certainty of reception and another one for larger ones which doesn't use additional buffers.

Reply:
Yes, you are correct the receiver will have to store the message somewhere. And without dynamic allocation, this is probably our best bet. I think we can have it like this for now? I am mostly worried about QMK ram usage. But if it does not pose a problem for now, I think it is acceptable

Reply:
I'll change max message size to 64 at least. The theoretical max for current messages should be around 57. I have C18 and tested only on this. 

ledMasks will take also around 280 bytes afaik. As transmission is on per-row basis it can be reduced too - the handling code will be more complicated though.

Reply:
I'd be glad if someone with C15 checked if after changes it works ok. before closing this conversation. Reduced stack size for the blinker thread, max message sizes, but kept the protocol and ledMasks intact for now.